File: adapters/aspnet/src/BugFab.AspNetCore/Storage/FileStorage.cs
using System.Collections.Concurrent;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace BugFab.AspNetCore.Storage;

/// <summary>
/// Disk-only <see cref="IStorage"/>. One folder per report, with
/// <c>metadata.json</c>, <c>lifecycle.json</c>, and <c>screenshot.png</c>
/// inside.
/// </summary>
/// <remarks>
/// Suitable for tests, demos, and single-process deployments. NOT
/// suitable for multi-replica production — there's no transactional locking.
/// Switch to <see cref="EfCoreStorage"/> for production.
/// </remarks>
public sealed class FileStorage : IStorage
{
    private readonly BugFabOptions _options;
    private readonly object _idLock = new();
    private readonly JsonSerializerOptions _jsonOptions;

    public FileStorage(BugFabOptions options)
    {
        _options = options;
        _jsonOptions = options.JsonOptions;
        Directory.CreateDirectory(options.StorageDirectory);
    }

    public Task<string> SaveReportAsync(
        JsonObject metadata,
        ReadOnlyMemory<byte> screenshot,
        CancellationToken ct = default)
    {
        var id = AllocateId();
        var dir = Path.Combine(_options.StorageDirectory, id);
        Directory.CreateDirectory(dir);

        var receivedAt = DateTimeOffset.UtcNow;
        metadata["id"] = id;
        metadata["created_at"] = receivedAt.ToString("O");
        metadata["status"] = "open";
        metadata["has_screenshot"] = true;

        File.WriteAllText(
            Path.Combine(dir, "metadata.json"),
            metadata.ToJsonString(_jsonOptions));

        var lifecycle = new JsonArray
        {
            new JsonObject
            {
                ["action"] = "created",
                ["by"] = (metadata["reporter"] as JsonObject)?["email"]?.GetValue<string>() ?? "anonymous",
                ["at"] = receivedAt.ToString("O"),
                ["status"] = "open",
            },
        };
        File.WriteAllText(
            Path.Combine(dir, "lifecycle.json"),
            lifecycle.ToJsonString(_jsonOptions));

        File.WriteAllBytes(Path.Combine(dir, "screenshot.png"), screenshot.ToArray());

        return Task.FromResult(id);
    }

    public Task<StoredReport?> GetReportAsync(string reportId, CancellationToken ct = default)
    {
        var dir = Path.Combine(_options.StorageDirectory, reportId);
        var metaPath = Path.Combine(dir, "metadata.json");
        if (!File.Exists(metaPath)) return Task.FromResult<StoredReport?>(null);

        var raw = JsonNode.Parse(File.ReadAllText(metaPath)) as JsonObject ?? new JsonObject();
        var lifecyclePath = Path.Combine(dir, "lifecycle.json");
        var lifecycle = File.Exists(lifecyclePath)
            ? (JsonNode.Parse(File.ReadAllText(lifecyclePath)) as JsonArray)?
                .OfType<JsonObject>().ToList() ?? new List<JsonObject>()
            : new List<JsonObject>();

        var ctxNode = raw["context"] as JsonObject ?? new JsonObject();
        var reporterNode = raw["reporter"] as JsonObject ?? new JsonObject();
        var tags = (raw["tags"] as JsonArray)?
            .Select(t => t?.GetValue<string>() ?? string.Empty)
            .Where(s => !string.IsNullOrEmpty(s))
            .ToList() ?? new List<string>();

        return Task.FromResult<StoredReport?>(new StoredReport(
            Id: reportId,
            Title: raw["title"]?.GetValue<string>() ?? string.Empty,
            ReportType: raw["report_type"]?.GetValue<string>() ?? "bug",
            Severity: raw["severity"]?.GetValue<string>() ?? "medium",
            Status: raw["status"]?.GetValue<string>() ?? "open",
            Module: ctxNode["module"]?.GetValue<string>() ?? string.Empty,
            CreatedAt: raw["created_at"]?.GetValue<string>() ?? string.Empty,
            UpdatedAt: raw["updated_at"]?.GetValue<string>() ?? string.Empty,
            HasScreenshot: File.Exists(Path.Combine(dir, "screenshot.png")),
            GitHubIssueUrl: raw["github_issue_url"]?.GetValue<string>(),
            GitHubIssueNumber: raw["github_issue_number"] is JsonValue n && n.TryGetValue<int>(out var num)
                ? num
                : null,
            Description: raw["description"]?.GetValue<string>() ?? string.Empty,
            ExpectedBehavior: raw["expected_behavior"]?.GetValue<string>() ?? string.Empty,
            Tags: tags,
            Reporter: reporterNode,
            Context: ctxNode,
            Lifecycle: lifecycle,
            ServerUserAgent: raw["server_user_agent"]?.GetValue<string>() ?? string.Empty,
            ClientReportedUserAgent: raw["client_reported_user_agent"]?.GetValue<string>() ?? string.Empty,
            Environment: raw["environment"]?.GetValue<string>() ?? string.Empty,
            ClientTs: raw["client_ts"]?.GetValue<string>() ?? string.Empty,
            ProtocolVersion: raw["protocol_version"]?.GetValue<string>() ?? "0.1",
            MetadataJson: raw.ToJsonString()));
    }

    public Task<(IReadOnlyList<StoredReportSummary> Items, int Total)> ListReportsAsync(
        ReportFilters filters,
        int page,
        int pageSize,
        CancellationToken ct = default)
    {
        var rows = EnumerateAllReports(filters)
            .OrderByDescending(s => s.CreatedAt, StringComparer.Ordinal)
            .ToList();
        var total = rows.Count;
        var paged = rows.Skip((page - 1) * pageSize).Take(pageSize).ToList();
        return Task.FromResult<(IReadOnlyList<StoredReportSummary>, int)>((paged, total));
    }

    public Task<string?> GetScreenshotPathAsync(string reportId, CancellationToken ct = default)
    {
        var path = Path.Combine(_options.StorageDirectory, reportId, "screenshot.png");
        return Task.FromResult(File.Exists(path) ? path : null);
    }

    public async Task<StoredReport?> UpdateStatusAsync(
        string reportId,
        string status,
        string fixCommit,
        string fixDescription,
        string by,
        CancellationToken ct = default)
    {
        var dir = Path.Combine(_options.StorageDirectory, reportId);
        var metaPath = Path.Combine(dir, "metadata.json");
        if (!File.Exists(metaPath)) return null;

        var raw = JsonNode.Parse(File.ReadAllText(metaPath)) as JsonObject ?? new JsonObject();
        raw["status"] = status;
        raw["updated_at"] = DateTimeOffset.UtcNow.ToString("O");
        File.WriteAllText(metaPath, raw.ToJsonString(_jsonOptions));

        var lifecyclePath = Path.Combine(dir, "lifecycle.json");
        var arr = File.Exists(lifecyclePath)
            ? JsonNode.Parse(File.ReadAllText(lifecyclePath)) as JsonArray ?? new JsonArray()
            : new JsonArray();

        arr.Add(new JsonObject
        {
            ["action"] = "status_changed",
            ["by"] = by,
            ["at"] = DateTimeOffset.UtcNow.ToString("O"),
            ["status"] = status,
            ["fix_commit"] = fixCommit,
            ["fix_description"] = fixDescription,
        });
        File.WriteAllText(lifecyclePath, arr.ToJsonString(_jsonOptions));

        return await GetReportAsync(reportId, ct).ConfigureAwait(false);
    }

    public Task<bool> DeleteReportAsync(string reportId, CancellationToken ct = default)
    {
        var dir = Path.Combine(_options.StorageDirectory, reportId);
        if (!Directory.Exists(dir)) return Task.FromResult(false);
        Directory.Delete(dir, recursive: true);
        return Task.FromResult(true);
    }

    public async Task<int> BulkCloseFixedAsync(string by, CancellationToken ct = default)
    {
        var fixedReports = EnumerateAllReports(new ReportFilters(Status: "fixed"))
            .Select(s => s.Id)
            .ToList();
        foreach (var id in fixedReports)
        {
            await UpdateStatusAsync(id, "closed", string.Empty, string.Empty, by, ct)
                .ConfigureAwait(false);
        }
        return fixedReports.Count;
    }

    public Task<int> BulkArchiveClosedAsync(CancellationToken ct = default)
    {
        var archiveDir = Path.Combine(_options.StorageDirectory, "archive");
        Directory.CreateDirectory(archiveDir);

        var closed = EnumerateAllReports(new ReportFilters(Status: "closed"))
            .Select(s => s.Id)
            .ToList();
        foreach (var id in closed)
        {
            var src = Path.Combine(_options.StorageDirectory, id);
            var dst = Path.Combine(archiveDir, id);
            if (Directory.Exists(dst)) Directory.Delete(dst, recursive: true);
            Directory.Move(src, dst);
        }
        return Task.FromResult(closed.Count);
    }

    public Task SetGitHubLinkAsync(
        string reportId,
        int issueNumber,
        string issueUrl,
        CancellationToken ct = default)
    {
        var metaPath = Path.Combine(_options.StorageDirectory, reportId, "metadata.json");
        if (!File.Exists(metaPath)) return Task.CompletedTask;
        var raw = JsonNode.Parse(File.ReadAllText(metaPath)) as JsonObject ?? new JsonObject();
        raw["github_issue_number"] = issueNumber;
        raw["github_issue_url"] = issueUrl;
        File.WriteAllText(metaPath, raw.ToJsonString(_jsonOptions));
        return Task.CompletedTask;
    }

    public Task<IReadOnlyDictionary<string, int>> ComputeStatsAsync(CancellationToken ct = default)
    {
        var result = new Dictionary<string, int>(StringComparer.Ordinal)
        {
            ["open"] = 0,
            ["investigating"] = 0,
            ["fixed"] = 0,
            ["closed"] = 0,
            ["total"] = 0,
        };
        foreach (var summary in EnumerateAllReports(new ReportFilters()))
        {
            if (result.ContainsKey(summary.Status)) result[summary.Status]++;
            result["total"]++;
        }
        return Task.FromResult<IReadOnlyDictionary<string, int>>(result);
    }

    private IEnumerable<StoredReportSummary> EnumerateAllReports(ReportFilters filters)
    {
        if (!Directory.Exists(_options.StorageDirectory)) yield break;

        foreach (var dir in Directory.EnumerateDirectories(_options.StorageDirectory))
        {
            var id = Path.GetFileName(dir);
            if (id is null || id.Equals("archive", StringComparison.Ordinal)) continue;
            if (!id.StartsWith("bug-", StringComparison.Ordinal)) continue;

            var metaPath = Path.Combine(dir, "metadata.json");
            if (!File.Exists(metaPath)) continue;

            JsonObject raw;
            try
            {
                raw = JsonNode.Parse(File.ReadAllText(metaPath)) as JsonObject ?? new JsonObject();
            }
            catch (JsonException)
            {
                continue;
            }

            var status = raw["status"]?.GetValue<string>() ?? "open";
            var severity = raw["severity"]?.GetValue<string>() ?? "medium";
            var module = (raw["context"] as JsonObject)?["module"]?.GetValue<string>() ?? string.Empty;
            var environment = raw["environment"]?.GetValue<string>() ?? string.Empty;

            if (!string.IsNullOrEmpty(filters.Status) && status != filters.Status) continue;
            if (!string.IsNullOrEmpty(filters.Severity) && severity != filters.Severity) continue;
            if (!string.IsNullOrEmpty(filters.Module) && module != filters.Module) continue;
            if (!string.IsNullOrEmpty(filters.Environment) && environment != filters.Environment) continue;

            yield return new StoredReportSummary(
                Id: id,
                Title: raw["title"]?.GetValue<string>() ?? string.Empty,
                ReportType: raw["report_type"]?.GetValue<string>() ?? "bug",
                Severity: severity,
                Status: status,
                Module: module,
                CreatedAt: raw["created_at"]?.GetValue<string>() ?? string.Empty,
                HasScreenshot: File.Exists(Path.Combine(dir, "screenshot.png")),
                GitHubIssueUrl: raw["github_issue_url"]?.GetValue<string>());
        }
    }

    private string AllocateId()
    {
        lock (_idLock)
        {
            int max = 0;
            if (Directory.Exists(_options.StorageDirectory))
            {
                foreach (var dir in Directory.EnumerateDirectories(_options.StorageDirectory))
                {
                    var name = Path.GetFileName(dir);
                    if (name is null || !name.StartsWith("bug-", StringComparison.Ordinal)) continue;
                    var digits = new string(name.Skip(4).SkipWhile(char.IsLetter).ToArray());
                    if (int.TryParse(digits, out var n) && n > max) max = n;
                }
            }
            var prefix = string.IsNullOrEmpty(_options.IdPrefix) ? string.Empty : _options.IdPrefix;
            return $"bug-{prefix}{max + 1:D3}";
        }
    }
}

File: adapters/express/src/storage/FileStorage.ts
// FileStorage — zero-dependency JSON-on-disk backend.
//
// Layout:
//   {storageDir}/{reportId}/metadata.json
//   {storageDir}/{reportId}/screenshot.png
//   {storageDir}/archive/{reportId}/...   (after bulk-archive-closed)
//
// Atomic metadata writes via temp-file-then-rename. Loads an in-memory
// index on construction so list/filter ops do not hit the filesystem.
//
// Multi-process caveat: the in-memory counter and index are not safe for
// concurrent multi-process writers. For PM2 / clustered deployments,
// implement a SQL-backed IStorage instead. See the README's "Custom
// backends" section for the IStorage contract.

import {
  copyFileSync, existsSync, mkdirSync, readdirSync, readFileSync,
  renameSync, rmSync, statSync, writeFileSync,
} from 'node:fs'
import { join, resolve } from 'node:path'
import { randomUUID } from 'node:crypto'

import type {
  IStorage, SaveReportInput, BugReportDetail, BugReportSummary,
  BugReportListStats, ListFilters, Status, LifecycleEvent, ReportType, Severity,
  Reporter, BugReportContext,
} from '../types.js'

interface StoredReport {
  id:                          string
  protocol_version:            string
  title:                       string
  description:                 string
  expected_behavior:           string
  report_type:                 ReportType
  severity:                    Severity
  status:                      Status
  tags:                        string[]
  reporter:                    Reporter
  context:                     BugReportContext
  client_ts:                   string
  server_user_agent:           string
  client_reported_user_agent:  string
  created_at:                  string
  updated_at:                  string
  archived_at:                 string | null
  github_issue_url:            string | null
  github_issue_number:         number | null
  lifecycle:                   LifecycleEvent[]
}

function nowIso(): string {
  return new Date().toISOString()
}

function pad(n: number, width = 3): string {
  return String(n).padStart(width, '0')
}

function summaryFrom(r: StoredReport): BugReportSummary {
  const summary: BugReportSummary = {
    id:               r.id,
    title:            r.title,
    report_type:      r.report_type,
    severity:         r.severity,
    status:           r.status,
    created_at:       r.created_at,
    has_screenshot:   true,
    github_issue_url: r.github_issue_url,
  }
  if (r.context?.module && typeof r.context.module === 'string') {
    summary.module = r.context.module
  }
  return summary
}

function detailFrom(r: StoredReport): BugReportDetail {
  return {
    id:                          r.id,
    title:                       r.title,
    report_type:                 r.report_type,
    severity:                    r.severity,
    status:                      r.status,
    module:                      typeof r.context?.module === 'string' ? r.context.module : '',
    created_at:                  r.created_at,
    has_screenshot:              true,
    github_issue_url:            r.github_issue_url,
    description:                 r.description,
    expected_behavior:           r.expected_behavior,
    tags:                        r.tags,
    reporter:                    r.reporter,
    context:                     r.context,
    lifecycle:                   r.lifecycle,
    server_user_agent:           r.server_user_agent,
    client_reported_user_agent:  r.client_reported_user_agent,
    environment:                 typeof r.context?.environment === 'string' ? r.context.environment : '',
    client_ts:                   r.client_ts,
    protocol_version:            r.protocol_version,
    updated_at:                  r.updated_at,
    github_issue_number:         r.github_issue_number,
  }
}

export interface FileStorageOptions {
  storageDir: string
  /** Optional ID prefix character — `"P"` produces ids like `bug-P001`. */
  idPrefix?:  string
}

export class FileStorage implements IStorage {
  private readonly dir:    string
  private readonly prefix: string
  private counter = 0
  // Active (non-archived) reports — backs `getReport`, default `listReports`.
  private index = new Map<string, StoredReport>()
  // Archived reports — surfaced only when `listReports` is called with
  // `include_archived=true`. Loaded from `<dir>/archive/<id>/metadata.json`
  // on construction and kept in sync by `archiveReport`. The archived map
  // is intentionally separate from the active map so default listings
  // never have to filter archived rows out and the archive view never
  // accidentally falls back into the live mutation paths.
  private archivedIndex = new Map<string, StoredReport>()

  constructor(opts: FileStorageOptions) {
    this.dir    = resolve(opts.storageDir)
    this.prefix = opts.idPrefix ?? ''
    mkdirSync(this.dir, { recursive: true })
    this.loadIndex()
  }

  private loadIndex(): void {
    let entries
    try {
      entries = readdirSync(this.dir, { withFileTypes: true })
    } catch {
      return
    }
    for (const entry of entries) {
      if (!entry.isDirectory()) continue
      if (entry.name === 'archive') {
        this.loadArchiveIndex()
        continue
      }
      const metaPath = join(this.dir, entry.name, 'metadata.json')
      if (!existsSync(metaPath)) continue
      try {
        const r = JSON.parse(readFileSync(metaPath, 'utf8')) as StoredReport
        this.index.set(r.id, r)
        this.bumpCounter(r.id)
      } catch {
        // Corrupt metadata file — skip and continue loading the rest.
      }
    }
  }

  /**
   * Walk `<dir>/archive/<id>/metadata.json` and populate `archivedIndex`.
   *
   * Counter bookkeeping must include archived ids too, otherwise a process
   * restart that finds only archived reports on disk would assign duplicate
   * ids on the next `saveReport` call.
   */
  private loadArchiveIndex(): void {
    const archiveRoot = join(this.dir, 'archive')
    if (!existsSync(archiveRoot)) return
    let archiveEntries
    try {
      archiveEntries = readdirSync(archiveRoot, { withFileTypes: true })
    } catch {
      return
    }
    for (const entry of archiveEntries) {
      if (!entry.isDirectory()) continue
      const metaPath = join(archiveRoot, entry.name, 'metadata.json')
      if (!existsSync(metaPath)) continue
      try {
        const r = JSON.parse(readFileSync(metaPath, 'utf8')) as StoredReport
        this.archivedIndex.set(r.id, r)
        this.bumpCounter(r.id)
      } catch {
        // Corrupt archived metadata — skip and continue.
      }
    }
  }

  private bumpCounter(id: string): void {
    const match = id.match(/(\d+)$/)
    if (match) {
      const n = parseInt(match[1]!, 10)
      if (n > this.counter) this.counter = n
    }
  }

  private nextId(): string {
    this.counter += 1
    const n = pad(this.counter, Math.max(3, String(this.counter).length))
    return this.prefix ? `bug-${this.prefix}${n}` : `bug-${n}`
  }

  private reportDir(id: string):       string { return join(this.dir, id) }
  private metaPath(id: string):        string { return join(this.reportDir(id), 'metadata.json') }
  private screenshotPath(id: string):  string { return join(this.reportDir(id), 'screenshot.png') }

  private writeMeta(report: StoredReport): void {
    const dest = this.metaPath(report.id)
    const tmp  = `${dest}.tmp-${randomUUID()}`
    writeFileSync(tmp, JSON.stringify(report, null, 2), 'utf8')
    renameSync(tmp, dest)
    this.index.set(report.id, report)
  }

  /** Returns an opaque file:// URI for the stored report directory. */
  storedAtFor(id: string): string {
    return `file://${this.reportDir(id).replace(/\\/g, '/')}/`
  }

  async saveReport(metadata: SaveReportInput, screenshotBytes: Buffer): Promise<string> {
    const id = this.nextId()
    const ts = nowIso()

    mkdirSync(this.reportDir(id), { recursive: true })
    writeFileSync(this.screenshotPath(id), screenshotBytes)

    const report: StoredReport = {
      id,
      protocol_version:            metadata.protocol_version,
      title:                       metadata.title,
      description:                 metadata.description,
      expected_behavior:           metadata.expected_behavior,
      report_type:                 metadata.report_type,
      severity:                    metadata.severity,
      status:                      'open',
      tags:                        metadata.tags,
      reporter:                    metadata.reporter,
      context:                     metadata.context,
      client_ts:                   metadata.client_ts,
      server_user_agent:           metadata.server_user_agent,
      client_reported_user_agent:  metadata.client_reported_user_agent,
      created_at:                  ts,
      updated_at:                  ts,
      archived_at:                 null,
      github_issue_url:            null,
      github_issue_number:         null,
      lifecycle: [{ action: 'created', by: 'anonymous', at: ts }],
    }

    this.writeMeta(report)
    return id
  }

  async getReport(id: string): Promise<BugReportDetail | null> {
    const r = this.index.get(id)
    return r ? detailFrom(r) : null
  }

  async listReports(
    filters:  ListFilters,
    page:     number,
    pageSize: number,
  ): Promise<{ items: BugReportSummary[]; total: number; stats: BugReportListStats }> {
    // include_archived=true merges the archive index into the visible set.
    // The active index never contains archived rows (they move to
    // archivedIndex on archiveReport) so the default listing has zero
    // overhead from the archive feature.
    const visible: StoredReport[] = filters.include_archived
      ? [...this.index.values(), ...this.archivedIndex.values()]
      : [...this.index.values()]

    let rows = visible
    if (filters.status)      rows = rows.filter((r) => r.status === filters.status)
    if (filters.severity)    rows = rows.filter((r) => r.severity === filters.severity)
    if (filters.environment) {
      rows = rows.filter((r) => {
        const env = r.context?.environment
        return typeof env === 'string' && env === filters.environment
      })
    }

    // Stats are computed across the filtered-but-pre-status-cut dataset so
    // the dashboard counters reflect "what would be visible given this set
    // of filters minus the status one." This matches the reference adapter.
    const statsBase = visible.filter((r) => {
      if (filters.severity && r.severity !== filters.severity) return false
      if (filters.environment) {
        const env = r.context?.environment
        if (typeof env !== 'string' || env !== filters.environment) return false
      }
      return true
    })
    const stats: BugReportListStats = {
      open:          statsBase.filter((r) => r.status === 'open').length,
      investigating: statsBase.filter((r) => r.status === 'investigating').length,
      fixed:         statsBase.filter((r) => r.status === 'fixed').length,
      closed:        statsBase.filter((r) => r.status === 'closed').length,
    }

    rows.sort((a, b) => b.created_at.localeCompare(a.created_at))
    const total = rows.length
    const items = rows
      .slice((page - 1) * pageSize, page * pageSize)
      .map(summaryFrom)

    return { items, total, stats }
  }

  async getScreenshotPath(id: string): Promise<string | null> {
    const p = this.screenshotPath(id)
    return existsSync(p) ? p : null
  }

  async updateStatus(
    id:              string,
    newStatus:       Status,
    by:              string,
    fixCommit?:      string,
    fixDescription?: string,
  ): Promise<BugReportDetail> {
    const r = this.index.get(id)
    if (!r) throw new Error(`Report not found: ${id}`)

    const at = nowIso()
    const entry: LifecycleEvent = {
      action: 'status_changed',
      by,
      at,
      status: newStatus,
    }
    if (fixCommit !== undefined)      entry.fix_commit      = fixCommit
    if (fixDescription !== undefined) entry.fix_description = fixDescription

    const updated: StoredReport = {
      ...r,
      status:     newStatus,
      updated_at: at,
      lifecycle:  [...r.lifecycle, entry],
    }

    this.writeMeta(updated)
    return detailFrom(updated)
  }

  async deleteReport(id: string): Promise<void> {
    if (!this.index.has(id)) throw new Error(`Report not found: ${id}`)
    const dir = this.reportDir(id)
    if (existsSync(dir)) rmSync(dir, { recursive: true, force: true })
    this.index.delete(id)
  }

  async archiveReport(id: string): Promise<void> {
    const r = this.index.get(id)
    if (!r) throw new Error(`Report not found: ${id}`)

    const archiveDir = join(this.dir, 'archive', id)
    mkdirSync(archiveDir, { recursive: true })

    const srcMeta = this.metaPath(id)
    const srcShot = this.screenshotPath(id)
    if (existsSync(srcMeta)) copyFileSync(srcMeta, join(archiveDir, 'metadata.json'))
    if (existsSync(srcShot)) copyFileSync(srcShot, join(archiveDir, 'screenshot.png'))

    // Apply archived_at + lifecycle entry, persist into the archive dir.
    const at = nowIso()
    const archived: StoredReport = {
      ...r,
      archived_at: at,
      lifecycle: [...r.lifecycle, { action: 'archived', by: 'system', at }],
    }
    const tmp = join(archiveDir, `metadata.json.tmp-${randomUUID()}`)
    writeFileSync(tmp, JSON.stringify(archived, null, 2), 'utf8')
    renameSync(tmp, join(archiveDir, 'metadata.json'))

    const srcDir = this.reportDir(id)
    if (existsSync(srcDir)) rmSync(srcDir, { recursive: true, force: true })
    this.index.delete(id)
    // Track the archived report so `listReports({ include_archived: true })`
    // can surface it. Without this, the on-disk archive folder would be the
    // only record and the JSON list endpoint would silently no-op.
    this.archivedIndex.set(id, archived)
  }

  async bulkCloseFixed(): Promise<number> {
    const fixed = [...this.index.values()].filter((r) => r.status === 'fixed' && !r.archived_at)
    for (const r of fixed) {
      await this.updateStatus(r.id, 'closed', 'system')
    }
    return fixed.length
  }

  async bulkArchiveClosed(): Promise<number> {
    const closed = [...this.index.values()].filter((r) => r.status === 'closed' && !r.archived_at)
    for (const r of closed) {
      await this.archiveReport(r.id)
    }
    return closed.length
  }

  async setGitHubIssue(id: string, issueUrl: string, issueNumber: number): Promise<void> {
    const r = this.index.get(id)
    if (!r) return
    const updated: StoredReport = {
      ...r,
      github_issue_url:    issueUrl,
      github_issue_number: issueNumber,
    }
    this.writeMeta(updated)
  }

  /** Test helper — number of currently-loaded reports. */
  size(): number {
    return this.index.size
  }

  /** Test helper — bytes written for a report's screenshot, or 0 if missing. */
  screenshotSize(id: string): number {
    const p = this.screenshotPath(id)
    if (!existsSync(p)) return 0
    return statSync(p).size
  }
}
